src/MongoDB.Driver/Linq/Linq3Implementation/Misc/MethodCallExpressionExtensions.cs

```diff
@@ -0,0 +1,37 @@
+﻿/* Copyright 2010-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+
+namespace MongoDB.Driver.Linq.Linq3Implementation.Misc
+{
+    internal static class MethodCallExpressionExtensions
+    {
+        public static Type GetResultType(this MethodCallExpression expression)
+        {
+            var resultType = expression.Type;
+            if (resultType.IsConstructedGenericType && resultType.GetGenericTypeDefinition() == typeof(Task<>))
+            {
+                return resultType.GetGenericArguments()[0];
+            }
+            else
+            {
+                return resultType;
+            }
+        }
```

Comment:
I really like this. It expresses intent well while encapsulating the details. Results in much cleaner code below. 👍